lib/rules/newline-before-return.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * @fileoverview Rule to check empty line before "return" statement
+ * @author Eric Clemmons <eric@smarterspam.com>
+ * @copyright 2014 Eric Clemmons. All rights reserved.
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = function(context) {
+
+    // Cache line numbers of comments for faster lookup
+    var comments = context.getAllComments().map(function (token) {
+        return token.loc.start.line;
+    });
+
+    //--------------------------------------------------------------------------
+    // Helpers
+    //--------------------------------------------------------------------------
+
+    function hasEmptyLineBefore(node) {
+        var previousToken = context.getTokenBefore(node);
+
+        var lineNum = node.loc.start.line,
+            previousLineNum = previousToken.loc.end.line;
+
+        // Measure the # of lines between tokens
+        var gap = lineNum - previousLineNum - 1;
+
+        // Comments don't count towards the gap
+        while (previousLineNum++ < lineNum) {
+            if (comments.indexOf(previousLineNum) !== -1) {
```

Comment:
Rather than searching through comments each time, you could just get the leading comments for the given node and test those `context.getComments(node).leading `. You can probably save a lot of  work that way.